guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+A function solves the problems that hooks solve without introducing the problems that they simultaneously introduce. With this pattern, tests that need to customize setup data can do so easily by passing options to the function, and this draws a clear line from the verification phase to the setup phase:
+
+✅
+
+``` typescript
+type WithControllerOptions = Partial<TokensControllerOptions>;
+
+type WithControllerCallback = (setup: {
+  controller: TokensController;
+  onNetworkDidChangeListener: (networkState: NetworkState) => void;
+}) => void | Promise<void>;
+
+type WithControllerArgs =
+  | [WithControllerOptions, WithControllerCallback]
+  | [WithControllerCallback];
+
+async function withController(...args: WithControllerArgs) {
+  const [{
+    chainId = '0x1',
+    onNetworkDidChange = (listener) => {
+      onNetworkDidChangeListener = listener;
+    },
+  }, fn] = args.length === 1 ? [{}, args[0]] : args;
+  const onNetworkDidChangeListener: (networkState: NetworkState) => void;
+
+  const controller = new TokensController({ chainId, onNetworkDidChange });
+  assert(onNetworkDidChangeListener, "onNetworkDidChangeListener was not set")
+
+  try {
+    await fn({ controller, onNetworkDidChangeListener });
+  } catch {
+    controller.destroy();
+  }
+}
+
+describe('TokensController', () => {
+  describe('addToken', () => {
+    it('registers the given token under the default network, assuming it has not changed yet', async () => {
+      await withController({ chainId: '0x1' }, ({ controller }) => {
+        controller.addToken({ symbol: 'DAI' });
+
+        expect(controller.state.tokens).toStrictEqual({
+          '0x1': {
+            symbol: 'DAI',
+          },
+        });
+      });
+    });
+
+    it('registers the given token under the current network, even after it changes', () => {
+      await withController(({ controller, onNetworkDidChangeListener }) => {
+        onNetworkDidChangeListener({ chainId: '0x42' });
+
+        controller.addToken({ symbol: 'DAI' });
+
+        expect(controller.state.tokens).toStrictEqual({
+          '0x42': {
+            symbol: 'DAI',
+          },
+        });
+      });
+    });
+  });
+});
+```
+
+The only case in which `beforeEach` or `afterEach` is acceptable is when using the Jest API, such as when setting fake timers:
```

Comment:
Nit: softer wording:
```suggestion
The only case in which `beforeEach` or `afterEach` is recommended is when using the Jest API, such as when setting fake timers:
```

Reply:
Actually.... is this recommended? `useFakeTimers` is similar to other setup steps, in that it might be useful for some steps but not others.

Reply:
Fair point. I included this exception because I've noticed that `useFakeTimers` tends to only be set once at the top of the file (because all of the tests rely on timers in some way) and I haven't run into a case where a single test needs something different. But, of course, it's always possible. In any case, this advice definitely conflicts with the advice given in the test isolation guidelines.

Should we recommend that people use a wrapper function instead?

``` typescript
function withFakeTimers(fn: () => void | Promise<void>) { 
  jest.useFakeTimers();

  try {
    await fn();
  } finally {
    jest.useRealTimers();
  }
}
```

Reply:
We have quite a few tests in core that use fake timers selectively (e.g. the TokenRatesController). 

But regarding your question, a `withFakeTiimers` wrapper definitely makes sense. But the pattern I've used in the past is to have global teardown but not setup. e.g. call `jest.useRealTimers()` in `afterEach` but leave it up to each test to decide whether to enable fake timers or not. Similar to our typical `sinon.restore()` strategy.

I guess anytime we are "resetting" the global state back to its initial state, using `afterEach` seems fair game and unlikely to cause maintenance issues.

Reply:
Maybe the whole section should be updated to suggest avoiding `beforeEach`, and not mention `afterEach`? I'm having trouble thinking of a scenario where `afterEach` has caused problems.

Reply:
I've revised this section to talk primarily about `beforeEach`. I do include a section on handling `afterEach` if necessary but hid that (along with the full explanation) behind a disclosure. Updated here: [1a8381b](https://github.com/MetaMask/contributor-docs/pull/38/commits/1a8381b901963910688228d92853b647f988e5e5)